extension/dynamicconfig/service/remoteconfig.go

```diff
@@ -0,0 +1,156 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+
+	res "github.com/open-telemetry/opentelemetry-proto/gen/go/resource/v1"
+	pb "github.com/vmingchen/opentelemetry-proto/gen/go/collector/dynamicconfig/v1"
+)
+
+type UpdateStrategy uint8
```

Comment:
Great question, clarifying comments added.

UpdateStrategy was introduced as an optimization. To reduce the number of times the backend has to ping the remote config service, the backend may cache the received `ConfigResponse` on a call to `GetFingerprint()`, then use the cached response on the subsequent call to `BuildConfigResponse()` if needed. This scheme is adopted if `UpdateStrategy = OnGetFingerprint`. The `GetFingerprint()` and `BuildConfigResponse()` calls happen in quick succession in the current implementation of the gRPC service, making it a desirable optimization.

Otherwise, if `UpdateStrategy=Default`, then a request is made to the upstream configuration service for both `GetFingerprint()` and `BuildConfigResponse()`. The strategy will remain `Default` unless otherwise specified.

`UpdateStrategy` overall is not a mechanism that is exposed to the end user. It's made public for the sake of extension, if other individuals would like to write their own version of the config service or backends on the collector, but implementation-wise, `UpdateStrategy` is hidden behind the private `backend` field, so it's operation isn't intended to be a critical detail to operating the dynamic config extension. In hindsight, however, it may be cleaner to make the mechanism private, hiding the implementation details for any user.